plugin/evm/block_sync.go

```
@@ -0,0 +1,127 @@
+package evm
+
+import (
+	"fmt"
+
+	"github.com/ava-labs/coreth/core/types"
+	"github.com/ava-labs/coreth/params"
+	"github.com/ava-labs/coreth/plugin/evm/atomic"
+	"github.com/ava-labs/libevm/common"
+	"github.com/ava-labs/libevm/core/rawdb"
+	"github.com/ava-labs/libevm/log"
+)
+
+func (b *Block) acceptDuringSync() error {
+	log.Debug("Accepting block during sync", "hash", b.ID(), "height", b.Height())
+	vm := b.vm
+
+	// Although returning an error from Accept is considered fatal, it is good
+	// practice to cleanup the batch we were modifying in the case of an error.
+	defer vm.versiondb.Abort()
+
+	if err := b.acceptAtomicOps(); err != nil {
+		return fmt.Errorf("failed to accept atomic ops during accept: %w", err)
+	}
+
+	// Write the canonical hash to the chain database
+	batch := vm.chaindb.NewBatch()
+	rawdb.WriteCanonicalHash(batch, b.ethBlock.Hash(), b.ethBlock.NumberU64())
+	if err := batch.Write(); err != nil {
+		return fmt.Errorf("failed to write block during sync: %w", err)
+	}
+
+	return nil
+}
+
+func (b *Block) rejectDuringSync() error {
+	var (
+		vm    = b.vm
+		block = b.ethBlock
+	)
+
+	// Because this will not be called during bootstrapping, this is guaranteed to be a valid atomic operation
+	atomicState, err := b.vm.atomicBackend.GetVerifiedAtomicState(common.Hash(b.ID()))
+	if err != nil {
+		// should never occur since [b] must be verified before calling Reject
+		log.Error("Should never happen because block must be verified before calling Reject", "block", b.ID(), "height", b.Height())
```

better to panic on unexpected conditions. (this is during sync, low risk)